### ISSUE TITLE
FocusNode exception

### ISSUE THREAD

**Issue by kika** [AbdulRahmanAlHamali/flutter_typeahead#72]
I'm on 1.2.0 and latest `dev` channel Flutter. My application has a `TabBar` with 3 tabs and when it starts it shows the first tab. `TypeAhead` is used only on the second tab, but if I switch from tab1 to tab3 I receive this:
```
flutter: ══╡ EXCEPTION CAUGHT BY SCHEDULER LIBRARY ╞═════════════════════════════════════════════════════════
flutter: The following assertion was thrown during a scheduler callback:
flutter: A FocusNode was used after being disposed.
flutter: Once you have called dispose() on a FocusNode, it can no longer be used.
flutter:
flutter: When the exception was thrown, this was the stack:
flutter: #0      ChangeNotifier._debugAssertNotDisposed.<anonymous closure> (package:flutter/src/foundation/change_notifier.dart:105:9)
flutter: #1      ChangeNotifier._debugAssertNotDisposed (package:flutter/src/foundation/change_notifier.dart:111:6)
flutter: #2      ChangeNotifier.addListener (package:flutter/src/foundation/change_notifier.dart:141:12)
flutter: #3      _TypeAheadFieldState.initState.<anonymous closure> (package:flutter_typeahead/flutter_typeahead.dart:747:32)
flutter: #4      _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1008:15)
flutter: #5      _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:956:9)
flutter: #6      _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:860:5)
flutter: #10     _invoke (dart:ui/hooks.dart:219:10)
flutter: #11     _drawFrame (dart:ui/hooks.dart:178:3)
flutter: (elided 3 frames from package dart:async)
flutter: ════════════════════════════════════════════════════════════════════════════════════════════════════
```

It doesn't happen if I switch to tab2 (where `TypeAhead` is used), only if I switch from tab1 to tab3 or from tab3 to tab1. I'm struggling to create a repro case though, but maybe someone has an idea what's going on?

**Comment by KaYBlitZ** [comment]
I don't know exactly how TabBar works, but it seems that somehow by switch tabs from 1 -> 3 or vice-versa, the TypeAheadField is being initialized and disposed quickly in rapid succession. By the time the next frame comes around and is invoked, the focus node is already disposed, causing the crash you see. The line causing the crash is:
```
this._effectiveFocusNode.addListener(_focusNodeListener);
```

I think a quick fix is wrap everything in a mounted check to see if the widget is still mounted before continuing like so:
```
WidgetsBinding.instance.addPostFrameCallback((duration) {
      if (mounted) { // ADD THIS LINE
        this._initOverlayEntry();
        // calculate initial suggestions list size
        this._suggestionsBoxController.resize();

        this._effectiveFocusNode.addListener(_focusNodeListener);

        // in case we already missed the focus event
        if (this._effectiveFocusNode.hasFocus) {
          this._suggestionsBoxController.open();
        }

        ScrollableState scrollableState = Scrollable.of(context);
        if (scrollableState != null) {
          // The TypeAheadField is inside a scrollable widget
          scrollableState.position.isScrollingNotifier.addListener(() {
            bool isScrolling = scrollableState.position.isScrollingNotifier
                .value;
            _resizeOnScrollTimer?.cancel();
            if (isScrolling) {
              // Scroll started
              _resizeOnScrollTimer =
                  Timer.periodic(_resizeOnScrollRefreshRate, (timer) {
                    _suggestionsBoxController.resize();
                  });
            } else {
              // Scroll finished
              _suggestionsBoxController.resize();
            }
          });
        }
      } // ADD THIS CLOSING BRACKET
    });
```
Can you try adding that quick check to see if the problem goes away?

**Comment by sjmcdowall** [comment]
FocusNodes are very odd things especially in a TabNav .. where all the tabs are under a single widget tree focus thing .. meaning the focus nodes all interact with each other and setting Focus in one can cause other focus things to not happen in another.  It's sort of a mess .. Need to make sure the "correct" focus Node is disposed .. and what is happening in the lifecycle of the widgets... 

I'd put print() statements in the initState() and dispose() of each widget in the NavBar and see what is happening (and I guess in the build()) .. so you can see the flow more easily of what happens..
 
A simple sample app that shows this behavior would also be good if you are still having issues ..

**Comment by KaYBlitZ** [comment]
@kika I am playing around and the example app included with this package has the exact same problem. I can confirm that switching between tabs 1 & 3 will cause the error. Furthermore, the `if(mounted)` check does fix this problem. We can add that in a future revision. It is strange why Flutter is creating and disposing of the 2nd tab when we are not even switching to it. Might be an issue to bring up with the Flutter team.

**Comment by sjmcdowall** [comment]
I'd like to see the code for the NavTab logic -- having implemented htis before I haven't seen this -- so I am wondering if there is some odd code in how the tabs are working... if you care to share ..

**Comment by KaYBlitZ** [comment]
The example project in TypeAhead has a TabBar. I used that to reproduce the problem.

**Comment by KaYBlitZ** [comment]
Fixed in #75 (version 1.2.1). Thanks for the bug report.

**Comment by kika** [comment]
Sorry for disappearing, I was flying around the country for work. I tried with the tip of the master and can confirm that it's fixed. Thanks so much!
